tensorpack/graph_builder/distributed.py

```diff
@@ -277,27 +290,101 @@ def build(self, get_grad_fn, get_opt_fn):
-            self._shadow_model_vars = DistributedReplicatedBuilder._shadow_model_variables(self._shadow_vars)
+        DataParallelBuilder._check_grad_list(self.grad_list)
+
+        aggregation_ops_list = []
+        if self.aggregation_frequency > 1:
+            assert not self.gpu_shadow_vars
+            # Create the variables which will be used to store aggregated updates.
+            for k, (grad_list, device) in enumerate(zip(self.grad_list, self.raw_devices)):
+                self.gpu_shadow_vars.append([])
+                with tf.device(device), tf.variable_scope("aggregation_variables"):
+                    for idx, (grad, var) in enumerate(grad_list):
+                        grad_aggregation_variable_name = self._get_grad_aggregator_variable_name(
+                            tower_id=k, variable_id=idx)
+                        grad_aggregation_variable = tf.get_variable(
+                            grad_aggregation_variable_name, shape=grad.get_shape().as_list(),
+                            trainable=False, initializer=tf.zeros_initializer(),
+                            collections=[tf.GraphKeys.LOCAL_VARIABLES, "aggregating_collection"])
+                        self.gpu_shadow_vars[k].append((grad_aggregation_variable, None))
+            assert len(self.gpu_shadow_vars) == len(self.grad_list)
+
+            # Apply new gradients to aggregation variables.
+            for k, (grad_list, device) in enumerate(zip(self.grad_list, self.raw_devices)):
+                with tf.device(device), tf.variable_scope("aggregation_variables", reuse=True):
+                    for idx, (grad, var) in enumerate(grad_list):
+                        grad_aggregation_variable_name = self._get_grad_aggregator_variable_name(
+                            tower_id=k, variable_id=idx)
+                        grad_aggregator = tf.get_variable(grad_aggregation_variable_name)
+                        aggregation_ops_list.append(grad_aggregator.assign_add(grad))
+                        self.gpu_shadow_vars[k][idx] = (self.gpu_shadow_vars[k][idx][0], var)
+        aggregation_ops = tf.group(*aggregation_ops_list)
+
+        # TODO: Use tf.queue instead of ConditionalAccumulator.
+        with tf.control_dependencies([aggregation_ops]):
+            # Using a large local step count so that it's always bigger than global step, which is incremented
+            # every time we call self.counter.take_grad().
+            train_op = self.counter.apply_grad(tf.constant([1], dtype=tf.float32), local_step=990000000)
+
+        # Communication op begins here.
+        ready_to_communicate = self.counter.take_grad(self.aggregation_frequency)
+
+        if self.aggregation_frequency > 1:
+            # Read in latest variables values.
+            aggregated_grads = []
+            aggregation_read_ops_list = []
+            with tf.control_dependencies([ready_to_communicate]):
+                for k,_ in enumerate(self.gpu_shadow_vars):
```

Comment:
`for k, (vars, device) in enumerate(zip(self.gpu_shadow_vars, self.raw_devices))`

Then replace `self.gpu_shadow_vars[k]` with `vars` and `self.raw_devices[k]` with `device`.

This should define a variable instead of `_` and use it below instead of reindexing into `gpu_shadow_vars`.